index.bs

```diff
@@ -1088,34 +1104,111 @@ strong map from handle ids to their corresponding objects.
 
 <pre class="cddl remote-cddl local-cddl">
 Handle = text;
+SharedId = text;
+</pre>
 
+<pre class="cddl remote-cddl">
 RemoteReference = {
+  RemoteObjectReference //
+  SharedReference
+}
+
+RemoteObjectReference = {
    handle: Handle,
+   ?sharedId: SharedId
+   Extensible
+}
+
+SharedReference = {
+   ?handle: Handle,
+   sharedId: SharedId
    Extensible
 }
 </pre>
 
 Issue: handle "stale object reference" case.
 
+Note: if the provided reference has both <code>handle</code> and
+<code>sharedId</code>, the algorithm will first try to resolve <code>handle</code>,
+then <code>sharedId</code>, and if non of that succeeded, return a error.
+
+### Deserialization ### {#data-types-reference-deserialization}
+
 <div algorithm>
 To <dfn>deserialize remote reference</dfn> given |realm| and |remote reference|:
 
-1. Let |handle id| be the value of the <code>handle</code> field of |remote reference|.
+1. Let |remote value| be <code>null</code>.
+
+1. If |remote reference| matches the <code>RemoteObjectReference</code>
+   production, set |remote value| to the result of
+   [=deserialize remote object reference=] given |realm| and |remote reference|.
+
+1. If |remote value| is <code>null</code> and |remote reference| matches the
+  <code>SharedReference</code> production, set |remote value| to the result of
+  [=trying=] to [=deserialize shared reference=] given |realm| and
+  |remote reference|.
+
+1. If |remote value| is <code>null</code>, return [=error=] with [=error code=]
+   [=invalid argument=].
+
+1. Return [=success=] with data |remote value|.
+
+</div>
+
+<div algorithm>
+To <dfn>deserialize remote object reference</dfn> given |realm| and
+|remote object reference|:
+
+1. Let |handle id| be the value of the <code>handle</code> field of
+   |remote object reference|.
 
 1. Let |handle map| be |realm|'s [=handle object map=]
 
-1. If |handle map| does not contain |handle id|, then return [=error=] with
-   [=error code=] [=invalid argument=].
+1. If |handle map| does not contain |handle id|, then return <code>null</code>.
 
 1. Return |handle map|[|handle id|].
 
 </div>
 
+<div algorithm>
+To <dfn>deserialize shared reference</dfn> given |realm| and |shared reference|:
+
+1. Let |browsing context| be [=get the browsing context=] of the |realm|.
+
+1. If |browsing context| is <code>null</code>, return [=error=] with [=error code=]
+   [=invalid argument=].
+
+1. Let |shared id| be the value of the <code>sharedId</code> field of
+   |shared reference|.
+
+1. If |shared id| is <code>null</code>, return [=error=] with [=error code=]
+   [=invalid argument=].
+
+1. Let |node| be result of [=trying=] to [=get a node=] given [=current session=],
+   |browsing context| and |shared id|.
+
+1. If |browsing context| is <code>null</code>, return [=error=] with [=error code=]
+   [=no such element=].
+
+1. If |node| [=is stale=], return [=error=] with [=error code=]
+   [=stale element reference=].
```

Comment:
As we want Classic to be implementable over BiDi, the `is stale` check is needed. I'm fine with throwing "no such element" instead.

Reply:
We shouldn't do either. We should put enough data on the serialized element that WebDriver classic implementations built on BiDi can implement this check.

However, what we probably do need to do here is check that `|node|`'s `ownerDocument`'s [origin](https://dom.spec.whatwg.org/#concept-document-origin) is [same origin-domain](https://html.spec.whatwg.org/#same-origin-domain) with `|context|`'s associated Document. 

@foolip to check the details here: what we're aiming for is that `|node|` is reachable from scripts running in `|realm|`.